File: ports/zig/src/widgets/shadow.zig
//! A drop shadow cast by a region onto whatever is behind it.
//!
//! The point is that it dims what it covers rather than painting over it: a
//! shadow that filled its band with a flat colour would erase the dashboard
//! underneath, which is the opposite of what a shadow is for. Every covered cell
//! keeps its own character and its own hue, and only loses some of its light.

const std = @import("std");

const buffer_mod = @import("../buffer.zig");
const color_mod = @import("../color.zig");
const layout = @import("../layout.zig");
const surface_mod = @import("../surface.zig");

const Color = color_mod.Color;
const Rect = layout.Rect;
const Style = buffer_mod.Style;
const Surface = surface_mod.Surface;

pub const ShadowOptions = struct {
    /// How far the shadow falls.
    offset_x: isize = 1,
    offset_y: isize = 1,
    /// 0-1: how much light the covered cells lose.
    amount: f64 = 0.55,
    /// Paint this colour instead of dimming what is underneath.
    ///
    /// For a shadow falling on empty background, where there is nothing to dim
    /// and a flat colour is cheaper and reads the same.
    color: ?Color = null,
};

/// Darken every cell in a region, keeping its character and its hue.
///
/// Towards black rather than towards the theme's background: on a light theme
/// the background *is* the light, so dimming towards it would make the shadow
/// brighter than the page it falls on.
pub fn dimRect(s: Surface, x: isize, y: isize, width: usize, height: usize, amount: f64) void {
    const t = std.math.clamp(amount, 0, 1);
    const black = Color.rgb(0, 0, 0);
    for (0..height) |row| {
        for (0..width) |col| {
            const ax = s.rect.x + x + @as(isize, @intCast(col));
            const ay = s.rect.y + y + @as(isize, @intCast(row));
            if (ax < s.clip.x or ay < s.clip.y or
                ax >= s.clip.x + @as(isize, @intCast(s.clip.width)) or
                ay >= s.clip.y + @as(isize, @intCast(s.clip.height))) continue;
            const i = s.buffer.index(@intCast(ax), @intCast(ay));
            s.buffer.fg[i] = s.buffer.fg[i].mix(black, t);
            s.buffer.bg[i] = s.buffer.bg[i].mix(black, t);
        }
    }
}

/// Cast a shadow from `rect` onto `s`.
///
/// The shadow is the band the region would cover if it were moved by the offset,
/// minus the region itself. Drawn before the region is, so it never falls on top
/// of it.
pub fn drawShadow(s: Surface, rect: Rect, options: ShadowOptions) void {
    if (s.isEmpty()) return;
    const dx = options.offset_x;
    const dy = options.offset_y;
    if (dx == 0 and dy == 0) return;

    const paint = struct {
        fn go(sf: Surface, o: ShadowOptions, x: isize, y: isize, w: isize, h: isize) void {
            if (w <= 0 or h <= 0) return;
            if (o.color) |c| {
                sf.fillRect(x, y, @intCast(w), @intCast(h), .{ .bg = c }, 32);
            } else {
                dimRect(sf, x, y, @intCast(w), @intCast(h), o.amount);
            }
        }
    }.go;

    // The shadow is the moved region minus the original, which splits into two
    // rectangles that do not touch: the rows the move added, at the moved
    // region's full width, and then the columns it added over the rows the two
    // still share. Cutting it any other way overlaps at the corner, and a corner
    // dimmed twice reads as a smudge rather than an edge.
    const rw: isize = @intCast(rect.width);
    const rh: isize = @intCast(rect.height);
    const tx = rect.x + dx;
    const ty = rect.y + dy;
    if (dy > 0) {
        paint(s, options, tx, rect.y + rh, rw, dy);
    } else if (dy < 0) {
        paint(s, options, tx, ty, rw, -dy);
    }

    const y0 = @max(rect.y, ty);
    const shared = @min(rect.y + rh, ty + rh) - y0;
    if (shared > 0) {
        if (dx > 0) {
            paint(s, options, rect.x + rw, y0, dx, shared);
        } else if (dx < 0) {
            paint(s, options, tx, y0, -dx, shared);
        }
    }
}

File: ports/rust/src/widgets/shadow.rs
//! A drop shadow cast by a region onto whatever is behind it.
//!
//! The point is that it dims what it covers rather than painting over it: a
//! shadow that filled its band with a flat colour would erase the dashboard
//! underneath, which is the opposite of what a shadow is for. Every covered cell
//! keeps its own character and its own hue, and only loses some of its light.

use crate::buffer::Style;
use crate::color::Color;
use crate::layout::Rect;
use crate::surface::Surface;

#[derive(Clone, Copy, Debug)]
pub struct ShadowOptions {
    /// How far the shadow falls. Default one cell right and one down.
    pub offset_x: isize,
    pub offset_y: isize,
    /// 0-1: how much light the covered cells lose.
    pub amount: f64,
    /// Paint this colour instead of dimming what is underneath.
    ///
    /// For a shadow falling on empty background, where there is nothing to dim
    /// and a flat colour is cheaper and reads the same.
    pub color: Option<Color>,
}

impl Default for ShadowOptions {
    fn default() -> ShadowOptions {
        ShadowOptions { offset_x: 1, offset_y: 1, amount: 0.55, color: None }
    }
}

/// Darken every cell in a region, keeping its character and its hue.
///
/// Towards black rather than towards the theme's background: on a light theme
/// the background *is* the light, so dimming towards it would make the shadow
/// brighter than the page it falls on.
pub fn dim_rect(surface: &Surface, x: isize, y: isize, width: usize, height: usize, amount: f64) {
    let t = amount.clamp(0.0, 1.0);
    let black = Color::rgb(0, 0, 0);
    for row in 0..height as isize {
        for col in 0..width as isize {
            let ax = surface.rect.x + x + col;
            let ay = surface.rect.y + y + row;
            if ax < surface.clip.x
                || ay < surface.clip.y
                || ax >= surface.clip.x + surface.clip.width as isize
                || ay >= surface.clip.y + surface.clip.height as isize
            {
                continue;
            }
            let mut buffer = surface.buffer_mut();
            let i = buffer.index(ax as usize, ay as usize);
            buffer.fg[i] = buffer.fg[i].mix(black, t);
            buffer.bg[i] = buffer.bg[i].mix(black, t);
        }
    }
}

/// Cast a shadow from `rect` onto `surface`.
///
/// The shadow is the band the region would cover if it were moved by the offset,
/// minus the region itself -- an L along the two trailing edges. Drawn before
/// the region is, so it never falls on top of it.
pub fn draw_shadow(surface: &Surface, rect: Rect, options: &ShadowOptions) {
    if surface.is_empty() {
        return;
    }
    let (dx, dy) = (options.offset_x, options.offset_y);
    if dx == 0 && dy == 0 {
        return;
    }

    let paint = |x: isize, y: isize, w: isize, h: isize| {
        if w <= 0 || h <= 0 {
            return;
        }
        match options.color {
            Some(color) => surface.fill_rect(
                x,
                y,
                w as usize,
                h as usize,
                &Style { fg: None, bg: Some(color), attrs: None },
                32,
            ),
            None => dim_rect(surface, x, y, w as usize, h as usize, options.amount),
        }
    };

    // The shadow is the moved region minus the original, which splits into two
    // rectangles that do not touch: the rows the move added, at the moved
    // region's full width, and then the columns it added over the rows the two
    // still share. Cutting it any other way overlaps at the corner, and a
    // corner dimmed twice reads as a smudge rather than an edge.
    let rw = rect.width as isize;
    let rh = rect.height as isize;
    let tx = rect.x + dx;
    let ty = rect.y + dy;
    if dy > 0 {
        paint(tx, rect.y + rh, rw, dy);
    } else if dy < 0 {
        paint(tx, ty, rw, -dy);
    }

    let y0 = rect.y.max(ty);
    let shared = (rect.y + rh).min(ty + rh) - y0;
    if shared > 0 {
        if dx > 0 {
            paint(rect.x + rw, y0, dx, shared);
        } else if dx < 0 {
            paint(tx, y0, -dx, shared);
        }
    }
}
